File: hackit/wireless/plugins/lua/scripts/eapol_analyzer.lua
local pcap_file = arg[1]

local function json_escape(s)
  return '"' .. tostring(s):gsub('\\', '\\\\'):gsub('"', '\\"'):gsub('\n', '\\n'):gsub('\r', '\\r') .. '"'
end

local function telemetry(event, data)
  local ts = tostring(os.time())
  io.write('{"event":' .. json_escape(event) .. ',"iface":"","bssid":"","ssid":"","channel":"","data":' .. data .. ',"timestamp":' .. json_escape(ts) .. '}\n')
  io.flush()
end

local handshake_steps = {[1] = 0, [2] = 0, [3] = 0, [4] = 0}
local anonce = nil
local snonce = nil
local mic_value = nil
local replay_counter = nil
local key_descriptors = {}

local function hex_bytes(s)
  if not s then return "" end
  local out = {}
  for i = 1, #s do
    out[#out + 1] = string.format("%02x", string.byte(s, i))
  end
  return table.concat(out, ":")
end

local function parse_eapol_key(data)
  if #data < 95 then return nil end
  local desc = {}
  desc.descriptor_type = string.byte(data, 1)
  if desc.descriptor_type ~= 2 then return nil end
  desc.key_info = string.byte(data, 2) + string.byte(data, 3) * 256
  desc.key_length = string.byte(data, 4) + string.byte(data, 5) * 256
  desc.replay_counter = string.byte(data, 6) + string.byte(data, 7) * 256 + string.byte(data, 8) * 65536 + string.byte(data, 9) * 16777216
  desc.nonce = data:sub(14, 45)
  desc.iv = data:sub(47, 62)
  desc.rsc = data:sub(63, 70)
  desc.key_id = data:sub(71, 72)
  desc.mic = data:sub(73, 88)
  desc.data_len = string.byte(data, 89) + string.byte(data, 90) * 256
  desc.key_data = data:sub(91, 91 + desc.data_len - 1)
  return desc
end

local function detect_handshake_step(eapol)
  local ki = eapol.key_info
  local install = (ki & 0x0040) ~= 0
  local ack = (ki & 0x0080) ~= 0
  local mic = (ki & 0x0100) ~= 0
  local secure = (ki & 0x0200) ~= 0
  local error_val = (ki & 0x0400) ~= 0
  local request = (ki & 0x0800) ~= 0

  local step = nil
  if ack and not mic and not install and not secure then
    step = 1
  elseif not ack and mic and not install and not secure then
    step = 2
  elseif ack and mic and install and not secure then
    step = 3
  elseif not ack and mic and not install and secure then
    step = 4
  end
  return step, ki
end

local function analyze_pcap()
  local cmd = "tshark -r " .. pcap_file .. " -Y \"eapol\" -T fields -e frame.number -e frame.time_epoch -e wlan.sa -e wlan.da -e eapol.keydes.keyinfo -e eapol.keydes.nonce -e eapol.keydes.mic -e eapol.keydes.replay_counter -e eapol.keydes.keydata 2>/dev/null"
  local f = io.popen(cmd)
  if not f then
    local cmd2 = "tcpdump -r " .. pcap_file .. " -xx -l -s 256 2>/dev/null | head -500"
    f = io.popen(cmd2)
    if not f then
      telemetry("error", '{"error":"Cannot read pcap file. Install tshark or tcpdump."}')
      return
    end
    telemetry("parse_mode", '{"method":"tcpdump_raw"}')
    local raw = f:read("*a")
    f:close()
    if #raw == 0 then
      telemetry("error", '{"error":"No data in pcap file"}')
      return
    end
    telemetry("raw_data", '{"length":' .. tostring(#raw) .. '}')
    return
  end

  telemetry("parse_mode", '{"method":"tshark"}')
  local frame_count = 0
  local total_eapol = 0

  while true do
    local line = f:read("*l")
    if not line then break end
    local parts = {}
    for val in line:gmatch("%S+") do
      parts[#parts + 1] = val
    end
    if #parts < 4 then
      telemetry("skip", '{"line":' .. json_escape(line:sub(1, 100)) .. '}')
    end
    total_eapol = total_eapol + 1
    local frame_num = parts[1] or "?"
    local timestamp = parts[2] or "0"
    local sa = parts[3] or "00:00:00:00:00:00"
    local da = parts[4] or "00:00:00:00:00:00"
    local keyinfo_hex = parts[5] or "0000"
    local nonce_hex = parts[6] or ""
    local mic_hex = parts[7] or ""
    local rc_val = parts[8] or "0"

    local ki = tonumber(keyinfo_hex, 16) or 0
    local nonce_bytes = ""
    for b in nonce_hex:gmatch("%x%x") do
      nonce_bytes = nonce_bytes .. string.char(tonumber(b, 16))
    end
    local mic_bytes = ""
    for b in mic_hex:gmatch("%x%x") do
      mic_bytes = mic_bytes .. string.char(tonumber(b, 16))
    end

    local step = nil
    local install = (ki & 0x0040) ~= 0
    local ack = (ki & 0x0080) ~= 0
    local mic_f = (ki & 0x0100) ~= 0
    local secure = (ki & 0x0200) ~= 0

    if ack and not mic_f and not install and not secure then step = 1 end
    if not ack and mic_f and not install and not secure then step = 2 end
    if ack and mic_f and install and not secure then step = 3 end
    if not ack and mic_f and not install and secure then step = 4 end

    if step then
      handshake_steps[step] = handshake_steps[step] + 1
      local nonce_str = ""
      if step == 1 then
        anonce = nonce_bytes
        nonce_str = hex_bytes(anonce)
      elseif step == 2 then
        snonce = nonce_bytes
        nonce_str = hex_bytes(snonce)
      end
      if #mic_hex > 0 then mic_value = mic_hex end
      replay_counter = rc_val

      telemetry("handshake_step", '{"step":' .. tostring(step) .. ',"frame":' .. json_escape(frame_num) .. ',"sa":' .. json_escape(sa) .. ',"da":' .. json_escape(da) .. ',"nonce":' .. json_escape(nonce_str) .. ',"mic":' .. json_escape(mic_hex) .. ',"replay_counter":' .. json_escape(tostring(rc_val)) .. ',"key_info_hex":' .. json_escape(keyinfo_hex) .. '}')
      frame_count = frame_count + 1

      if step == 4 then
        local integrity = "valid"
        if mic_hex == "00000000000000000000000000000000" then
          integrity = "invalid_zero_mic"
        end
        telemetry("handshake_complete", '{"total_steps":4,"frames":' .. tostring(frame_count) .. ',"step_counts":{"1":' .. tostring(handshake_steps[1]) .. ',"2":' .. tostring(handshake_steps[2]) .. ',"3":' .. tostring(handshake_steps[3]) .. ',"4":' .. tostring(handshake_steps[4]) .. '},"integrity":' .. json_escape(integrity) .. '}')
      end
    else
      telemetry("eapol_other", '{"frame":' .. json_escape(frame_num) .. ',"key_info":' .. json_escape(keyinfo_hex) .. ',"type":"non_handshake_or_unknown"}')
    end
  end
  f:close()

  telemetry("eapol_summary", '{"total_eapol":' .. tostring(total_eapol) .. ',"handshake_frames":' .. tostring(frame_count) .. ',"steps_found":{"1":' .. tostring(handshake_steps[1]) .. ',"2":' .. tostring(handshake_steps[2]) .. ',"3":' .. tostring(handshake_steps[3]) .. ',"4":' .. tostring(handshake_steps[4]) .. '}}')
end

if not pcap_file then
  telemetry("error", '{"error":"Usage: eapol_analyzer.lua <pcap_file>"}')
  os.exit(1)
end

local f = io.open(pcap_file, "r")
if not f then
  telemetry("error", '{"error":"Cannot open pcap file: ' .. pcap_file .. '"}')
  os.exit(1)
end
f:close()

local ok, err = pcall(analyze_pcap)
if not ok then
  telemetry("error", '{"error":"Analysis failed: ' .. json_escape(tostring(err)) .. '"}')
  os.exit(1)
end

telemetry("complete", '{"file":' .. json_escape(pcap_file) .. '}')

File: hackit/wireless/plugins/ruby/scripts/eapol_analyzer.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require 'json'
require 'time'

def emit(event, iface, bssid, ssid, channel, data)
  puts JSON.generate(event: event, iface: iface, bssid: bssid, ssid: ssid, channel: channel, data: data, timestamp: Time.now.iso8601)
  $stdout.flush
end

def mac_str(bytes)
  bytes.unpack('H2 H2 H2 H2 H2 H2').map { |b| b.upcase }.join(':')
end

EAPOL_HEADER_SIZE = 4

# EAPOL types
EAPOL_EAP = 0
EAPOL_START = 1
EAPOL_LOGOFF = 2
EAPOL_KEY = 3
EAPOL_ENCAP_ASF = 4

EAPOL_TYPES = %w[EAP EAPOL-Start EAPOL-Logoff EAPOL-Key Encapsulated-ASF].freeze

# Key Descriptor Versions
KEY_DESC_V1 = 1
KEY_DESC_V2 = 2

KEY_INFO_ACK = 0x0080
KEY_INFO_MIC = 0x0100
KEY_INFO_SECURE = 0x0200
KEY_INFO_ERROR = 0x0400
KEY_INFO_REQUEST = 0x0800
KEY_INFO_INSTALL = 0x0040
KEY_INFO_PAIRWISE = 0x1000

def key_info_description(ki)
  parts = []
  parts << 'ACK' if ki & KEY_INFO_ACK != 0
  parts << 'MIC' if ki & KEY_INFO_MIC != 0
  parts << 'SECURE' if ki & KEY_INFO_SECURE != 0
  parts << 'ERROR' if ki & KEY_INFO_ERROR != 0
  parts << 'REQUEST' if ki & KEY_INFO_REQUEST != 0
  parts << 'INSTALL' if ki & KEY_INFO_INSTALL != 0
  parts << 'PAIRWISE' if ki & KEY_INFO_PAIRWISE != 0
  parts.join('|')
end

def analyze_eapol_key(key_data, bssid, client)
  return unless key_data.size >= 95

  desc_version = key_data[0].ord
  key_info = key_data[1..2].unpack1('v')
  key_len = key_data[3..4].unpack1('v')
  replay_counter = key_data[5..12].unpack1('Q>')
  nonce = key_data[13..44]
  iv = key_data[45..60]
  rsc = key_data[61..68]
  mic = key_data[77..92]
  # data_len follows at 93-94, then key_data at 95+

  handshake_msg = case [key_info & KEY_INFO_ACK != 0, key_info & KEY_INFO_MIC != 0, key_info & KEY_INFO_SECURE != 0, key_info & KEY_INFO_INSTALL != 0]
                  when [true, false, false, true] then 'Message 1'
                  when [false, true, false, false] then 'Message 2'
                  when [true, true, true, true] then 'Message 3'
                  when [false, true, true, false] then 'Message 4'
                  else 'Unknown'
                  end

  result = {
    desc_version: desc_version,
    key_info: format('0x%04x', key_info),
    key_info_desc: key_info_description(key_info),
    key_len: key_len,
    replay_counter: replay_counter,
    anonce: nonce.unpack1('H*'),
    snonce: nonce.unpack1('H*'),
    iv: iv.unpack1('H*'),
    rsc: rsc.unpack1('H*'),
    mic: mic.unpack1('H*'),
    handshake_msg: handshake_msg,
    pairwise: key_info & KEY_INFO_PAIRWISE != 0
  }

  emit('eapol_key_analysis', '', bssid, '', 0, result.merge(client: client))
  result
end

def process_pcap(pcap_path)
  unless File.exist?(pcap_path)
    emit('pcap_not_found', '', '', '', 0, { path: pcap_path })
    return
  end

  file_size = File.size(pcap_path)
  emit('pcap_open', '', '', '', 0, { path: pcap_path, size: file_size })

  eapol_frames = []
  handshake_complete = false
  handshake_params = {}

  File.open(pcap_path, 'rb') do |f|
    global_header = f.read(24)
    return unless global_header && global_header.size == 24

    magic = global_header[0..3]
    if magic == "\xa1\xb2\xc3\xd4" || magic == "\xd4\xc3\xb2\xa1"
      endian = magic == "\xd4\xc3\xb2\xa1" ? :little : :big
      emit('pcap_format', '', '', '', 0, { magic: magic.unpack1('H*'), endian: endian })
    else
      emit('pcap_invalid_magic', '', '', '', 0, { magic: magic.unpack1('H*') })
      return
    end

    max_frames = 10_000
    frame_count = 0

    while frame_count < max_frames
      pkt_header = f.read(16)
      break unless pkt_header && pkt_header.size == 16

      if endian == :little
        ts_sec = pkt_header[0..3].unpack1('V')
        incl_len = pkt_header[8..11].unpack1('V')
      else
        ts_sec = pkt_header[0..3].unpack1('N')
        incl_len = pkt_header[8..11].unpack1('N')
      end

      pkt_data = f.read(incl_len)
      break unless pkt_data && pkt_data.size == incl_len
      frame_count += 1

      radiotap_len = pkt_data[2..3].unpack1('v') rescue 0
      next if radiotap_len < 4 || radiotap_len >= pkt_data.size
      frame = pkt_data[radiotap_len..]
      next if frame.nil? || frame.size < 24

      fc = frame[0..1].unpack1('v')
      frame_type = (fc >> 2) & 0x3
      frame_subtype = (fc >> 4) & 0xF
      to_ds = (fc >> 8) & 0x1
      from_ds = (fc >> 9) & 0x1

      addr1 = mac_str(frame[4..9])
      addr2 = mac_str(frame[10..15])
      addr3 = mac_str(frame[16..21])
      bssid = if to_ds == 0 && from_ds == 0
                addr3
              elsif to_ds == 1 && from_ds == 0
                addr1
              elsif to_ds == 0 && from_ds == 1
                addr2
              else
                addr3
              end

      # Look for QoS + LLC + SNAP + EAPOL
      llc_offset = 24
      if (frame_type == 2 || frame_type == 0) && frame.size > llc_offset + 8
        snap = frame[llc_offset, 8]
        if snap && snap[0..2] == "\xaa\xaa\x03" && snap[3..4] == "\x00\x00" && snap[5..7] == "\x88\x8e\x47"
          eapol_start = llc_offset + 8
          eapol_pkt = frame[eapol_start..]
          next if eapol_pkt.nil? || eapol_pkt.size < EAPOL_HEADER_SIZE

          eapol_version = eapol_pkt[0].ord
          eapol_type = eapol_pkt[1].ord
          eapol_body_len = eapol_pkt[2..3].unpack1('n')
          body = eapol_pkt[4..] || ''

          next unless eapol_type == EAPOL_KEY && body.size >= 95

          key_data = body[0..94]
          client = if from_ds == 1 then addr2 else addr1 end

          analysis = analyze_eapol_key(key_data, bssid, client)
          eapol_frames << analysis

          case analysis[:handshake_msg]
          when 'Message 1'
            handshake_params[:anonce] = analysis[:anonce]
            handshake_params[:bssid] = bssid
            handshake_params[:client] = client
          when 'Message 2'
            handshake_params[:snonce] = analysis[:snonce] if handshake_params[:anonce]
          when 'Message 3'
            handshake_params[:anonce_m3] = analysis[:anonce]
            handshake_params[:mic_m3] = analysis[:mic]
          when 'Message 4'
            handshake_params[:mic_m4] = analysis[:mic]
            handshake_complete = true
          end

          if handshake_complete && handshake_params[:anonce] && handshake_params[:snonce]
            emit('handshake_complete', '', bssid, '', 0, handshake_params)
            handshake_complete = false
          end
        end
      end
    end

    emit('eapol_summary', '', '', '', 0, {
      total_eapol_frames: eapol_frames.size,
      frames_processed: frame_count,
      bssids: eapol_frames.map { |e| e[:bssid] }.compact.uniq
    })
  end

  emit('pcap_analysis_complete', '', '', '', 0, { path: pcap_path, eapol_frames: eapol_frames.size })
end

begin
  pcap_file = ARGV[0]
  raise 'pcap_file path required' unless pcap_file

  emit('eapol_analyzer_start', '', '', '', 0, { pcap: pcap_file })
  process_pcap(pcap_file)
  emit('eapol_analyzer_complete', '', '', '', 0, {})

rescue Interrupt
  emit('eapol_analyzer_interrupted', '', '', '', 0, {})
rescue => e
  $stderr.puts JSON.generate(event: 'error', data: { message: e.message, backtrace: e.backtrace&.first }, timestamp: Time.now.iso8601)
  exit 1
end
